### DISCUSSION TITLE
2024-10-18 - Pull Requests - Merge Debt - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#9436] [Chewing the Fat]
https://www.ssw.com.au/rules/merge-debt/

From your experience, what is the most common cause of Merge Debt?
What is the best technique or tool for reducing Merge Debt?

**Comment by tkapa** [comment]
> From your experience, what is the most common cause of Merge Debt?

Typically, it's comments awaiting resolution, or reviewers not generally looking at the PRs for their repo.

> What is the best technique or tool for reducing Merge Debt?

Keeping PRs smaller in general. It can be difficult to split a large PR into different PRs, but keeping those PRs small makes them digestible and easy to review.

**Comment by ozdemir-mehmet** [comment]
Busy reviewers, devs that don't followup

**Comment by edkranz** [comment]
1. Long-lived feature branches that have not been synchronised with main regularly.
2. `git merge main`, and for reviewers: only block a PR if absolutely necessary.

**Comment by calumjs** [comment]
>From your experience, what is the most common cause of Merge Debt?

PRs not getting reviewed due to complexity or lack of reviewer availability

>What is the best technique or tool for reducing Merge Debt?

Automate as much as possible

**Comment by bahjatalaadelssw** [comment]
> From your experience, what is the most common cause of Merge Debt?
No available reviewers

> What is the best technique or tool for reducing Merge Debt?
Having a pool of reviewers you're collected that become your unblockers (like a masters list)

**Comment by TobyChurches** [comment]
> From your experience, what is the most common cause of Merge Debt?

Comments asking for too big of a change (Exactly what was mentioned in the video) or Async collaboration (can slow down the process when the reviewer is in a different time zone)

> What is the best technique or tool for reducing Merge Debt?

Keep PR's small and comments relevant

**Comment by matt-goldman-ssw** [comment]
>From your experience, what is the most common cause of Merge Debt?

A combination of PRs being too big and requested changes being too big.

>What is the best technique or tool for reducing Merge Debt?

Keep PRs small and focused. If there's going to be a bigger PR it should be discussed and acknowledged, and planned. Then when ready the team can review together. Keeping the PR small will inherently keep the feedback small.

**Comment by awaltos** [comment]
> From your experience, what is the most common cause of Merge Debt?

Authors not following up their PRs. Reviewers requesting unnecessary changes

> What is the best technique or tool for reducing Merge Debt?

Keep PRs small - adhere to SRP.
Increasing visibility of open PRs. Having multiple people able to review across multiple teams

**Comment by RobJT-wq** [comment]
> 1. First think about something you learnt and what you think about the rules

I didn't know about Merge Debt prior to this CTF.

> 2. From your experience, what is the most common cause of Merge Debt?

I assume the most common cause would be PR's that are too large or complex and that the reviewer is unable to contact the colleague who created the PR

> What is the best technique or tool for reducing Merge Debt?

The best technique would be to ensure that a PR has been reviewed before working on the next PBI

**Comment by ozair-ashfaque** [comment]
> From your experience, what is the most common cause of Merge Debt?

Comments to be resolved or the reviewers are too busy reviewing the PR.

> What is the best technique or tool for reducing Merge Debt?

Small PRs

**Comment by vladislav-kir** [comment]
> From your experience, what is the most common cause of Merge Debt?

Reviewers are busy and Authors are not following up.

> What is the best technique or tool for reducing Merge Debt?

Keeping PRs small and easy to read.

**Comment by StefStar** [comment]
Can't speak for merge debt as not a dev but clear communication I think can address most cases of "debt" in any team

**Comment by danielmackay** [comment]
> From your experience, what is the most common cause of Merge Debt?

Developers overloading PR's with changes from multiple PBIs, or submitting a single massive PR instead of breaking it down to smaller PRs.

> What is the best technique or tool for reducing Merge Debt?

Get PR's reviewed quickly so that developers are encouraged to create smaller more frequent PRs.

**Comment by chrisschultzssw** [comment]
> From your experience, what is the most common cause of Merge Debt?

Reviewers being busy

> What is the best technique or tool for reducing Merge Debt?

TOFU

**Comment by bettybondoc** [comment]
> From your experience, what is the most common cause of Merge Debt?

No one available to review

> What is the best technique or tool for reducing Merge Debt?

TOFU

**Comment by Ben0189** [comment]
> Most common cause of Merge Debt?

Dev does not actively follow up their pr

> What is the best technique or tool for reducing Merge Debt?

Keep pr small, call reviewer and go through the pr together

**Comment by steven0x51** [comment]
> From your experience, what is the most common cause of Merge Debt?

PRs not reviewed in time.

> What is the best technique or tool for reducing Merge Debt?

Small PRs

**Comment by Pravino24** [comment]
It would be good talk with your PR approver to ensure the PR is as per requirement and so it gets approved without delay.

**Comment by ricksu978** [comment]
> From your experience, what is the most common cause of Merge Debt?

Huge PR with little unit testing

> What is the best technique or tool for reducing Merge Debt?

Breakdown PBI and keep PR small

**Comment by AntPolkanov** [comment]
> From your experience, what is the most common cause of Merge Debt?

Many updated files - difficult to review, difficult to test and even more difficult to review 2nd or 3rd round.

> What is the best technique or tool for reducing Merge Debt?

Small PRs

**Reply by matt-goldman-ssw** [reply]
> even more difficult to review 2nd or 3rd round.

This is so true. Especially on AzDo.

**Comment by Tanya-Leahy** [comment]
>From your experience, what is the most common cause of Merge Debt?

PRs not reviewed quick enough.

>What is the best technique or tool for reducing Merge Debt?

Small PRs with only changes to 1 file.

**Comment by jeoffreyfischer** [comment]
> From your experience, what is the most common cause of Merge Debt?

PRs not reviewed quick enough or reviewers approving PRs without giving feedback

> What is the best technique or tool for reducing Merge Debt?

Small PRs that fix one single problem

**Comment by JackDevAU** [comment]
Great video as always @matt-goldman-ssw & @lukecookssw 🚀

> From your experience, what is the most common cause of Merge Debt?

PRs left unreviewed for a while + Reviewers feedback left sitting for a few days.

> What is the best technique or tool for reducing Merge Debt?

Small PRs always can be merged quickly. But I think having confidence also plays a huge part, if the project has some great tests you can feel safer knowing its gonna be alright 😂

**Comment by sam-s-maher** [comment]
> From your experience, what is the most common cause of Merge Debt?

Hanging PR's, which leads to mover overhead as developers context switch between this PR and other PR's they are working on/reviewing. Even worse if the requirements change, scope increases, or other PBI's end up conflicting with it. 

> What is the best technique or tool for reducing Merge Debt?

The team committing to reviewing PR's quickly, and not getting caught in the rip-tide of multiple PR's. Extra points for developers thoroughly self reviewing their own PR's (which I can be guilty of not doing), as often a small typo that might take the author 10 minutes to review and fix can lead to a whole extra review cycle.

**Comment by uly1** [comment]
PRs are very annoying for content changes and they slow things down unneccessarily. They should be removed from Tina and rules, especially for the owners of these sites

**Reply by SamBWagner** [reply]
Agree. Content, in my opinion, should be super fast to change with little to no approval step. It's content at the end of the day and, so long as you're not giving access to bad actors, is very low risk.

**Comment by AttackOnMorty** [comment]
> What is the best technique or tool for reducing Merge Debt?

Raise small scoped PRs and set a timebox to follow up.

**Comment by amankumarrr** [comment]
> From your experience, what is the most common cause of Merge Debt?

Big PRs often take time and can lead to Merge debt

> What is the best technique or tool for reducing Merge Debt?

Split the PR if it’s getting too large.

**Comment by MattParkerDev** [comment]
> From your experience, what is the most common cause of Merge Debt?

PRs being too big or complicated, and reviewers avoiding these PRs. 

> What is the best technique or tool for reducing Merge Debt?

Keep PRs small and easy to be reviewed.

**Reply by SamBWagner** [reply]
100%! lil' PRs are easy to review **and** you end up getting a higher quality review in the end.

**Comment by stephanfakossw** [comment]
> From your experience, what is the most common cause of Merge Debt?

Waiting for review...

> What is the best technique or tool for reducing Merge Debt?

KISS

**Comment by SamBWagner** [comment]
> From your experience, what is the most common cause of Merge Debt?

Mixture of people just throwing a Pull Request up and then complaining that the reviewers aren't telepathic and didn't review their code instantly.

and

Requesters making extremely large pull requests and then getting sad/mad when people aren't reviewing it in a timely matter.

If your PR is huge, that's your fault, if your not directly asking someone to review your code, your basically just asking to be ignored. Just talk to people. 

> What is the best technique or tool for reducing Merge Debt?

Call people, message people, take the responsibility to make sure your code is reviewed and gets in. Your goal is to get your work done, if that's by pestering people then that's what needs to be done. I've seldom had people block a Pull Request over menial style choices. I might be lucky in that respect.

**Comment by babakamyljanovssw** [comment]
> From your experience, what is the most common cause of Merge Debt?

Big PRs

> What is the best technique or tool for reducing Merge Debt?

Relevant comments, keeping PRs small, if it is complex then call reviewer and do over shoulder review

**Comment by jimmidier** [comment]
> From your experience, what is the most common cause of Merge Debt?

Large PR is usually the cause, which is mostly originated from a large PBI that could be split into multiple sub-PBIs.

> What is the best technique or tool for reducing Merge Debt?

Make smaller commits and make sure each commit is runnable can help breaking large PR in the above case.

**Comment by PennyWalker** [comment]
It's hard for me to comment on as a non-dev, but from what I've observed, large PRs that have taken a long time to submit will create big problems. Busy reviewers taking a long time to review it.

**Comment by jerryluossw** [comment]
> From your experience, what is the most common cause of Merge Debt?

Reviewers are not available or PR is too complex.

> What is the best technique or tool for reducing Merge Debt?

Use AI code review tools to help reviewers understand the code in the PR more easily.

**Comment by bradystroud** [comment]
> From your experience, what is the most common cause of Merge Debt?

♻️ Big PRs create scared reviewers, scared reviewers create slow reviews, slow reviews create big PRs
 

> What is the best technique or tool for reducing Merge Debt?

Little PRs and TOFU and Dory

**Comment by Aibono1225** [comment]
> From your experience, what is the most common cause of Merge Debt?

Wait for the reviewers

> What is the best technique or tool for reducing Merge Debt?

Create small PRs, small commits. Call people

**Comment by miichaelsmedley** [comment]
> https://www.ssw.com.au/rules/merge-debt/
> 
> From your experience, what is the most common cause of Merge Debt? What is the best technique or tool for reducing Merge Debt?

Limit the PRs - keep them simple.

**Comment by tiagov8** [comment]
> From your experience, what is the most common cause of Merge Debt?

Dynamic teams, where devs are moved from one project to another end up making it harder to chase people up and also leave some projects overlooked. 

> What is the best technique or tool for reducing Merge Debt?

Taking responsibility to review PRs daily (I do on SSW.Rules, and try on SSW.People)

**Comment by Calinator444** [comment]
> From your experience, what is the most common cause of Merge Debt?

Busy developers. People don't like context switching to review code changes. Particularly large code changes.

> What is the best technique or tool for reducing Merge Debt?

Keeping your PRs small, easy to review, and pinging developers shortly after creating a PR. If the changes are complex ask the developers if they want to do an over the shoulder review.

**Comment by leahy268** [comment]
>From your experience, what is the most common cause of Merge Debt?

A Large PR is the worst.  When someone has edited 20 files it's really hard to be sure it's ok.  Especially with what I usually do which is IAC and Git changes

>What is the best technique or tool for reducing Merge Debt?

Keep the PR's small, then ping someone to get it done.  If they are unsure then call them and go through it.

**Comment by sethdaily** [comment]
> From your experience, what is the most common cause of Merge Debt?

People not taking responsibility for their PR

> What is the best technique or tool for reducing Merge Debt?

Always get your PR approved ASAP

**Comment by ncn-ssw** [comment]
> From your experience, what is the most common cause of Merge Debt?

Potential reviewers being too timid to make a firm 👍 or 👎 decision. I believe this typically arises due to a lack of experience.

> What is the best technique or tool for reducing Merge Debt?

- I believe the senior members of any team need to prioritize approving PRs quickly
- Over time, the junior members need to be encouraged to review PRs first to exercise their judgement - they can then check their comments with that of their more senior colleagues

As a result of this, members of the team will often start reviewing PRs with minimal prompting.

**Comment by wicksipedia** [comment]
> From your experience, what is the most common cause of Merge Debt?

no TOFU

> What is the best technique or tool for reducing Merge Debt?

Moar TOFU also the team can follow up on open PRs in the Daily Scrum

**Comment by drwharris** [comment]
> From your experience, what is the most common cause of Merge Debt?

Large, unthought through PRs

> What is the best technique or tool for reducing Merge Debt?

Thinking your changes through before you start them.
